### ISSUE TITLE
Error: Dependency is not defined

### ISSUE THREAD

**Issue by j4zzcat** [ets-labs/python-dependency-injector#198]
I wonder why the following code fails?

```python
from   collections import namedtuple
import dependency_injector.containers  as containers
import dependency_injector.providers   as providers

app_Config                = namedtuple( "Config",          [ "dir" ] )
app_ExternalConfigAdapter = namedtuple( "ConfigAdapter",   [ "config", "starting_point" ] )
external_RestApiClient    = namedtuple( "RestApiClient",   [ "config", "version" ] )

class CommonServices( containers.DeclarativeContainer ):
    config = providers.Singleton( app_Config, dir = "/root/.qrsa" )

class ExternalServices( containers.DeclarativeContainer ):
    common_services = providers.DependenciesContainer()

    adapted_config = providers.Factory( app_ExternalConfigAdapter,
        common_services.config,
        "common/nitro"
    )

    external_rest_client = providers.Dependency( instance_of = external_RestApiClient )
    external_rest_client.provided_by(
        providers.Singleton( external_RestApiClient,
            config  = adapted_config,
            version = "9.0"
        )
    )

cs  = CommonServices()
qrs = ExternalServices( common_services = cs )
qrs.adapted_config()

# --- fails here ---
qrs.external_rest_client()

```

**Comment by rmk135** [comment]
Hi @j4zzcat ,

I will do some debugging later today.

Roman

**Comment by rmk135** [comment]
Well, I think that a way to get it working is next:

```python
from   collections import namedtuple
import dependency_injector.containers  as containers
import dependency_injector.providers   as providers

app_Config                = namedtuple( "Config",          [ "dir" ] )
app_ExternalConfigAdapter = namedtuple( "ConfigAdapter",   [ "config", "starting_point" ] )
external_RestApiClient    = namedtuple( "RestApiClient",   [ "config", "version" ] )

class CommonServices( containers.DeclarativeContainer ):
    config = providers.Singleton( app_Config, dir = "/root/.qrsa" )

class ExternalServices( containers.DeclarativeContainer ):
    version = providers.Dependency()
    common_services = providers.DependenciesContainer()

    adapted_config = providers.Factory( app_ExternalConfigAdapter,
        common_services.config,
        "common/nitro"
    )

    external_rest_client = providers.Singleton( external_RestApiClient,
        config  = adapted_config,
        version = version
    )

cs  = CommonServices()
qrs = ExternalServices( common_services = cs , version = "9.0" )
qrs.adapted_config()

print(qrs.external_rest_client())
# RestApiClient(config=ConfigAdapter(config=Config(dir='/root/.qrsa'), starting_point='common/nitro'), version='9.0')
```

I will also check the behavior of `Dependency` provider in such case because, generally, your code looks "legal", but I need to debug more to check why it fails - there could be few reasons.

**Comment by rmk135** [comment]
@j4zzcat , I've found an issue, thank you very-very much. It was a really "nasty" bug :)

I will publish a fix in few hours or so.

**Comment by rmk135** [comment]
Hi @j4zzcat,

Issue has been fixed and new release [3.12.3](https://pypi.org/project/dependency_injector/3.12.3/) has been published to PyPi.

U're welcome to upgrade.

Thanks again for reporting the issue!

Regards,
Roman
